drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1901 @@
+		if (!data_buf || max_data_len != 8) {
+			LOG_ERR("Invalid parameter for reading Setup packet");
+			rc = -EINVAL;
+			goto cleanup;
+		}
+
+		memcpy(data_buf, &ep_mgmt->setup_packet, 8);
+		ep_mgmt->new_setup = false;
+
+		if (read_bytes) {
+			*read_bytes = 8;
+		}
+
+		LOG_DBG("ep_addr=0x%02x, read setup packet=8 bytes", ep_addr);
+
+		goto cleanup;
+	}
+
+	/* For one-shot implementation, don't read FIFO with EP busy. */
+	if (ep_cur->nak_clr) {
+		LOG_WRN("ep_addr 0x%02x busy", ep_addr);
+		rc = -EAGAIN;
+		goto cleanup;
+	}
+
+	/* NOTE: Null data and zero data length is valid, used for returning number of
+	 * available bytes for read
+	 */
+	if (data_buf) {
+		data_len_act = max_data_len;
+		rc = numaker_usbd_ep_fifo_copy_to_user(ep_cur, data_buf, &data_len_act);
+		if (rc < 0) {
+			LOG_ERR("Copy from FIFO to user buffer");
+			goto cleanup;
+		}
+
+		if (read_bytes) {
+			*read_bytes = data_len_act;
+		}
+	} else if (max_data_len) {
+		LOG_ERR("Null data but non-zero data length");
+		rc = -EINVAL;
+		goto cleanup;
+	} else {
+		if (read_bytes) {
+			*read_bytes = numaker_usbd_ep_fifo_used(ep_cur);
+		}
+	}
+
+	/* Suppress further USB_DC_EP_DATA_OUT events by replying NAK or disabling interrupt
+	 *
+	 * For USBD, further control is unnecessary because NAK is automatically replied until
+	 * next USBD_SET_PAYLOAD_LEN().
+	 */
+
+	LOG_DBG("ep_addr=0x%02x, read=%d bytes", ep_addr, data_len_act);
+
+cleanup:
+
+	m_numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+int usb_dc_ep_read_continue(uint8_t ep_addr)
+{
+	LOG_DBG("ep_addr=0x%02x", ep_addr);
+
+	struct usb_dc_numaker_device *dev = usb_dc_numaker_device_get();
+	int rc = 0;
+
+	m_numaker_usbd_lock(dev);
+
+	/* Bind EP context to EP address */
+	struct numaker_usbd_ep *ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep_addr);
+
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep_addr=0x%02x", ep_addr);
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+	if (!USB_EP_DIR_IS_OUT(ep_addr)) {
+		LOG_ERR("Invalid EP address 0x%02x for read", ep_addr);
+		rc = -EINVAL;
+		goto cleanup;
+	}
+
+	/* Avoid duplicate NAK clear */
+	if (ep_cur->nak_clr) {
+		LOG_DBG("ep_addr 0x%02x has triggered", ep_addr);
+		rc = 0;
+		goto cleanup;
+	}
+
+	/* For one-shot implementation, don't trigger next DATA OUT, or overwrite. */
+	if (numaker_usbd_ep_fifo_used(ep_cur)) {
+		goto cleanup;
+	}
+
+	__ASSERT_NO_MSG(ep_cur->ep_mps_valid);
+	numaker_usbd_ep_trigger(ep_cur, ep_cur->ep_mps);
+
+cleanup:
+
+	m_numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+int usb_dc_ep_mps(const uint8_t ep_addr)
+{
+	LOG_DBG("ep_addr=0x%02x", ep_addr);
+
+	struct usb_dc_numaker_device *dev = usb_dc_numaker_device_get();
+	int rc = 0;
+	uint16_t ep_mps = 0;
+
+	m_numaker_usbd_lock(dev);
+
+	/* Bind EP context to EP address */
+	struct numaker_usbd_ep *ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep_addr);
+
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep_addr=0x%02x", ep_addr);
+		rc = -ENOMEM;
+		goto cleanup;
```

Comment:
you are setting `rc` but tthen return `ep_mps`, pick one :-)

Reply:
yes, fixed